ASoC: SDCA: support Q7.8 volume format

shumingfan opened the pull request:
The SDCA specification uses Q7.8 volume format.
This patch adds a field to indicate whether it is SDCA volume control and supports the volume settings.

changed log
v2: create sdca_soc_get_volsw/sdca_soc_put_volsw in the SDCA code. For SDCA volume control only.
v3: use a function pointer for ctl_to_reg/reg_to_ctl for Q78 format

charleskeepax commented on include/sound/soc.h:
```diff
@@ -555,6 +555,7 @@ int snd_soc_get_volsw(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_elem_value *ucontrol);
 int snd_soc_put_volsw(struct snd_kcontrol *kcontrol,
 	struct snd_ctl_elem_value *ucontrol);
+
```
Probably shouldn't be changing whitespace in unrelated places.

shumingfan replied:
will fix

charleskeepax commented on include/sound/soc.h:
```diff
@@ -1225,6 +1226,7 @@ struct soc_mixer_control {
 	unsigned int sign_bit;
 	unsigned int invert:1;
 	unsigned int autodisable:1;
+	unsigned int sdca_vol_q7p8:1;
```
I am a little nervous about pulling this into the ASoC core, have you looked at adding wrappers for snd_soc_get_volsw/snd_soc_put_volsw in the SDCA code, if it doesn't end up looking too messy I feel like that might be preferrable.

shumingfan replied:
Understood. I will update the patch.

charleskeepax commented on sound/soc/sdca/sdca_asoc.c:
```diff
@@ -814,37 +813,23 @@ static int control_limit_kctl(struct device *dev,
+	mc->min = ((int)tlv[2] / step);
+	mc->max = ((int)tlv[3] / step);
+	mc->shift = step;
+	mc->rshift = step;
```
Are you sure you can just use step as the shift here? I don't think this works, step is the gap between valid values, but shift is an actual bit shift.

shumingfan replied:
The shift will be used in the soc_mixer_reg_to_ctl/soc_mixer_ctl_to_reg function. 
Do you think I should create another field to store the step value?

charleskeepax commented on include/sound/soc.h:
```diff
@@ -1220,6 +1220,7 @@ struct soc_mixer_control {
 	/* Limited maximum value specified as presented through the control */
 	int platform_max;
 	int reg, rreg;
+	int step;
```
If we are duplicating the handlers we can probably reuse the shift member for step, rather than adding a new one. I was just unsure before as we were using shift as a shift sometimes and as a step other times.

charleskeepax commented on sound/soc/sdca/sdca_asoc.c:
```diff
@@ -786,6 +786,126 @@ int sdca_asoc_populate_dapm(struct device *dev, struct sdca_function_data *funct
+	val -= mc->min;
+
+	if (mc->invert)
+		val = max - val;
```
Can probably drop the invert we dont use it for volume controls in SDCA.

charleskeepax commented on sound/soc/sdca/sdca_asoc.c:
```diff
@@ -786,6 +786,126 @@ int sdca_asoc_populate_dapm(struct device *dev, struct sdca_function_data *funct
+		return -EINVAL;
+
+	if (mc->sign_bit)
+		val = sign_extend32(val, mc->sign_bit);
```
q7.8 will always have a sign bit so can probably drop the if here.

charleskeepax commented on sound/soc/sdca/sdca_asoc.c:
```diff
@@ -786,6 +786,126 @@ int sdca_asoc_populate_dapm(struct device *dev, struct sdca_function_data *funct
 }
 EXPORT_SYMBOL_NS(sdca_asoc_populate_dapm, "SND_SOC_SDCA");
 
+static int sdca_soc_vol_reg_to_ctl(struct soc_mixer_control *mc, unsigned int reg_val,
```
Maybe update the name to include q78 or something incase we need to as most custom handlers in the future.

charleskeepax commented on sound/soc/soc-ops.c:
```diff
@@ -202,27 +231,35 @@ static int soc_put_volsw(struct snd_kcontrol *kcontrol,
-			val1 |= soc_mixer_ctl_to_reg(mc,
+			val1 |= ctl_to_reg(mc,
 						     ucontrol->value.integer.value[1],
 						     mask, mc->rshift, max);
```
We should sort the indentation on these.

shumingfan replied:
will fix

charleskeepax commented on sound/soc/soc-ops.c:
```diff
@@ -110,6 +110,35 @@ int snd_soc_put_enum_double(struct snd_kcontrol *kcontrol,
+		return -EINVAL;
+
+	reg_val = val + mc->min;
+	ret_val = (((int)(((int)reg_val * (int)mc->shift) << 8)) / 100);
```
Is all the casting necessary?

shumingfan replied:
will reduce that

charleskeepax commented:
I think this looks like it should work, still a little nervous about pushing changes into soc-ops for SDCA I am not sure it has the re-usability. I will have a poke see what it would look like if we pulled that in, but perhaps we should do the next bit of the review upstream see if Mark or anyone has any thoughts on that.

shumingfan committed:
ASoC: SDCA: support Q7.8 volume format

The SDCA specification uses Q7.8 volume format.
This patch adds a field to indicate whether it is SDCA volume control
and supports the volume settings.

Signed-off-by: Shuming Fan <shumingf@realtek.com>

shumingfan commented:
> I think this looks like it should work, still a little nervous about pushing changes into soc-ops for SDCA I am not sure it has the re-usability. I will have a poke see what it would look like if we pulled that in, but perhaps we should do the next bit of the review upstream see if Mark or anyone has any thoughts on that.

@charleskeepax Thanks for the review. Will send the patch and see if Mark or anyone has any suggestions.

bardliao commented:
The PR is upstreamed
